### PR TITLE
Refactor Level System

### PR DESCRIPTION
**Pull request by hyv1001**
Refactor Level System

- Using smart pointer to manage resources in `WorldManager` and `Level`
- Using `GObjectID ` as object id instead of `size_t` 
- Using ObjectIDAllocator to allocate id 
- Just load the default level instead of all levels
- Using character name as the current character key instead of the index

### CODE REVIEW

File: engine/source/runtime/function/framework/level/level.cpp [hunk 1]
```diff
@@ -18,73 +18,85 @@ namespace Pilot
...
-            GObject* character_object  = m_gobjects[level_res.m_character_index];
-            m_current_active_character = new Character(character_object);
+            std::shared_ptr<GObject> object = object_pair.second;
+            if (object == nullptr)
+                continue;
+
+            if (level_res.m_character_name == object->getName())
+            {
+                m_current_active_character = std::make_shared<Character>(object);
+				break;
```
Comment (Ol6rin) [inline engine/source/runtime/function/framework/level/level.cpp:80]: weird indent

File: engine/source/runtime/function/framework/object/object.cpp [hunk 2]
```diff
@@ -13,6 +13,8 @@
 
 namespace Pilot
 {
+    std::atomic<GObjectID> ObjectIDAllocator::m_next_id {0};
```
Comment (Ol6rin) [inline engine/source/runtime/function/framework/object/object.cpp:16]: Should be in object_id_allocator.cpp.

File: engine/source/runtime/function/framework/object/object_allocator.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    using GObjectID = size_t;
+
+    constexpr GObjectID k_invalid_go_id = std::numeric_limits<size_t>::max();
+
+    class ObjectIDAllocator
+    {
+    public:
+        static GObjectID alloc()
+        {
+            std::atomic<GObjectID> ret = m_next_id.load();
```
Comment (Ol6rin) [inline engine/source/runtime/function/framework/object/object_allocator.h:18]: Maybe ret should have a more meaningful name

File: engine/source/runtime/function/framework/object/object_id_allocator.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#pragma once
```
Comment (Ol6rin) [inline engine/source/runtime/function/framework/object/object_id_allocator.h:1]: Uh oh, duplicated files. object_allocator.h and object_id_allocator.h

File: engine/source/runtime/function/framework/object/object_id_allocator.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+namespace Pilot
+{
+    using GObjectID = size_t;
+
+    constexpr GObjectID k_invalid_go_id = std::numeric_limits<size_t>::max();
+
+    class ObjectIDAllocator
+    {
+    public:
+        static GObjectID alloc()
```
Comment (Ol6rin) [inline engine/source/runtime/function/framework/object/object_id_allocator.h:16]: Move implementation to cpp

File: engine/source/runtime/function/framework/world/world_manager.cpp [hunk 6]
```diff
@@ -1,121 +1,137 @@
...
     {
-        processPendingLoadWorld();
+        if (!m_is_world_loaded)
+        {
+            loadWorld(m_current_world_url);
+        }
 
-        for (Level* level : m_levels)
+        // tick the active level
+        auto active_level = m_current_active_level.lock();
```
Comment (Ol6rin) [inline engine/source/runtime/function/framework/world/world_manager.cpp:45]: Avoid auto here

File: engine/source/runtime/function/framework/world/world_manager.cpp [hunk 7]
```diff
@@ -1,121 +1,137 @@
...
+            return false;
+        }
 
-    void WorldManager::loadWorld(const std::string& world_url)
-    {
-        WorldRes world_res;
-        AssetManager::getInstance().loadAsset(world_url, world_res);
+        // set the default level to be active level
+        auto iter = m_loaded_levels.find(world_res.m_default_level_url);
+        ASSERT(iter != m_loaded_levels.end());
```
Comment (Ol6rin) [inline engine/source/runtime/function/framework/world/world_manager.cpp:72]: As we don't assert on resource error, we should LOG_FATAL here

File: engine/source/runtime/function/framework/world/world_manager.h [hunk 8]
```diff
@@ -11,6 +11,8 @@ namespace Pilot
     class PWorldT;
```
Comment (Ol6rin) [inline engine/source/runtime/function/framework/world/world_manager.h:11]: can be removed

File: engine/source/runtime/function/framework/world/world_manager.h [hunk 9]
```diff
@@ -23,26 +25,24 @@ namespace Pilot
...
-        std::string m_current_world_url;
-        std::string m_current_level_url;
+        bool                      m_is_world_loaded {false};
+        std::string               m_current_world_url;
+        std::shared_ptr<WorldRes> m_current_world_resource;
 
-        std::vector<Level*> m_levels;
-        Level*              m_current_active_level {nullptr};
+        // all loaded levels, key: level url, vaule: level instance
+        std::unordered_map<std::string, std::shared_ptr<Level>> m_loaded_levels;
```
Comment (Ol6rin) [inline engine/source/runtime/function/framework/world/world_manager.h:44]: typedef

**Review by Ol6rin — CHANGES_REQUESTED** [review]

File: engine/source/runtime/function/character/character.cpp [hunk 10]
```diff
@@ -7,9 +7,19 @@
 
 namespace Pilot
 {
-    Character::Character(GObject* character_object) { setObject(character_object); }
+    Character::Character(std::shared_ptr<GObject> character_object) { setObject(character_object); }
```
[the remark this answers is not in the document]
Comment (hyv1001) [reply engine/source/runtime/function/character/character.cpp:10]: I think the Character and Level both have the object's ownership in the current implementation. 
Suppose the object is deleted at level, how to tell the character. It's inappropriate to check the character's object while removing an object in the level.

[the remark this answers is not in the document]
Comment (Ol6rin) [reply engine/source/runtime/function/character/character.cpp:10]: We have done that in Level::deleteGObjectByID.
We checked whether the deleting gobject was the character object. If it was, we set the corresponding gobject in the character to nullptr.

File: engine/source/runtime/function/framework/world/world_manager.cpp (hunk 7 above)
Comment (hyv1001) [reply engine/source/runtime/function/framework/world/world_manager.cpp:72]: The logic of the above code will guarantee that the default level should be in loaded levels. This ASSERT is used for readiblity

**Pull request merged**